data/storage/etcd/store.go

```diff
@@ -149,44 +144,63 @@ func (s *etcd) Get(ctx context.Context, key string, out proto.Message, ignoreNot
 }
 
 // Update implements storage.Interface.Update
-func (s *etcd) Update(ctx context.Context, key string, val proto.Message, ttl int64) error {
+func (s *etcd) Update(ctx context.Context, key string, uf storage.UpdateFunc, template proto.Message) error {
 	if !s.connected {
 		if err := s.Connect(); err != nil {
 			return err
 		}
 	}
 	key = s.prefix(key)
 
-	// must exist
-	_, err := s.client.KV.Get(ctx, key)
+	// current state
+	getResp, err := s.client.KV.Get(ctx, key)
 	if err != nil {
 		return err
 	}
-
-	opts, err := s.options(ctx, int64(ttl))
-	if err != nil {
-		return err
+	if len(getResp.Kvs) == 0 {
+		return storage.NotFound
 	}
+	current := getResp.Kvs[0]
 
-	data, err := proto.Marshal(val)
+	// Begin update loop
+	for {
+		input := proto.Clone(template)
```

Comment:
It isn't immediately clear to me why this needs to be done each loop iteration. A comment would help...